ledger/eval/eval_test.go

```diff
@@ -1357,9 +1363,167 @@ func TestExpiredAccountGenerationWithDiskFailure(t *testing.T) {
 	require.ErrorContains(t, err, "disk I/O fail (on purpose)")
 }
 
+// TestAbsenteeChecks tests that the eval function will correctly mark accounts as absent
+func TestAbsenteeChecks(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+
+	genesisInitState, addrs, keys := ledgertesting.GenesisWithProto(10, protocol.ConsensusFuture)
+
+	// add 32 more addresses, we can get a suspension by challenge
+	for i := 0; i < 32; i++ {
+		addrs = append(addrs, basics.Address{byte(i << 3), 0xaa})
+	}
+
+	// Set all addrs to online
+	for i, addr := range addrs {
+		tmp := genesisInitState.Accounts[addr]
+		tmp.Status = basics.Online
+		crypto.RandBytes(tmp.StateProofID[:])
+		crypto.RandBytes(tmp.SelectionID[:])
+		crypto.RandBytes(tmp.VoteID[:])
+		tmp.VoteFirstValid = 1
+		tmp.VoteLastValid = 1500 // large enough to avoid EXPIRATION, so we can see SUSPENSION
+		tmp.LastHeartbeat = 1    // non-zero allows suspensions
+		switch i {
+		case 1:
+			tmp.LastHeartbeat = 1150 // lie hear so that addr[1] won't be suspended
+		case 2:
+			tmp.LastProposed = 1150 // lie hear so that addr[2] won't be suspended
+		}
+
+		genesisInitState.Accounts[addr] = tmp
+	}
+
+	l := newTestLedger(t, bookkeeping.GenesisBalances{
+		Balances:    genesisInitState.Accounts,
+		FeeSink:     testSinkAddr,
+		RewardsPool: testPoolAddr,
+	})
+
+	newBlock := bookkeeping.MakeBlock(l.blocks[0].BlockHeader)
+
+	blkEval, err := l.StartEvaluator(newBlock.BlockHeader, 0, 0, nil)
+	require.NoError(t, err)
+
+	// Advance the evaluator, watching for lack of suspensions since we don't
+	// suspend until a txn with a suspendable account appears
+	challenge := byte(0)
+	for i := uint64(0); i < uint64(1210); i++ { // A bit past one grace period (200) past challenge at 1000.
+		vb := l.endBlock(t, blkEval)
+		blkEval = l.nextBlock(t)
+		require.Zero(t, vb.Block().AbsentParticipationAccounts)
+		if vb.Block().Round() == 1000 {
+			challenge = vb.Block().BlockHeader.Seed[0]
+		}
+	}
+	challenged := basics.Address{(challenge >> 3) << 3, 0xaa}
```

Comment:
why do the shifts here?

Reply:
I pulled out the entire challenge byte, but only the top 5 bits matter. I created 32 addresses so that one of them would match those five top bits. Those addresses are all of the form `{0xXX, 0xaa}` where 0xXX has has a number from 0-31 in the top bits, and zeros in the low 3 bits. So once I know the challenge, I know which of my 32 addresses was challenged.  Line 1420 is that address. (I have to clear out those bottom 3 bits with the shifts in order to be the exact address I made.)

Reply:
Which means I could have done `& 0b11111000` which would probably be clearer. I forgot Go had binary constants and didn't think `& 0xf7` would have been as clear.